iis-wf/iis-wf-referenceextraction/src/main/java/eu/dnetlib/iis/wf/referenceextraction/patent/OpenPatentWebServiceFacade.java

```diff
@@ -0,0 +1,327 @@
+    private transient HttpHost opsHost;
+    
+    /**
+     * Serialization / deserialization details.
+     */
+    private SerDe serDe;
+    
+    
+    // ------------------- CONSTRUCTORS ------------------------
+    
+    public OpenPatentWebServiceFacade(int connectionTimeout, int readTimeout, 
+            String authHostName, int authPort, String authScheme, String authUriRoot,
+            String opsHostName, int opsPort, String opsScheme, String opsUriRoot, 
+            String consumerCredential, long throttleSleepTime, int maxRetriesCount) {
+        
+        this(buildHttpClient(connectionTimeout, readTimeout),
+                new HttpHost(authHostName, authPort, authScheme), authUriRoot, 
+                new HttpHost(opsHostName, opsPort, opsScheme), opsUriRoot,
+                consumerCredential, throttleSleepTime, maxRetriesCount, new JsonParser());
+        
+        // persisting for further serialization and deserialization
+        this.serDe = new SerDe(connectionTimeout, readTimeout, 
+                authHostName, authPort, authScheme, 
+                opsHostName, opsPort, opsScheme);
+    }
+    
+    /**
+     * Using this constructor simplifies object instantiation but also makes the whole object non-serializable.
+     */
+    protected OpenPatentWebServiceFacade(HttpClient httpClient, HttpHost authHost, String authUriRoot, HttpHost opsHost,
+            String opsUriRoot, String consumerCredential, long throttleSleepTime, int maxRetriesCount,
+            JsonParser jsonParser) {
+        this.httpClient = httpClient;
+        this.authHost = authHost;
+        this.authUriRoot = authUriRoot;
+        this.opsHost = opsHost;
+        this.opsUriRoot = opsUriRoot;
+        this.consumerCredential = consumerCredential;
+        this.throttleSleepTime = throttleSleepTime;
+        this.maxRetriesCount = maxRetriesCount;
+        this.jsonParser = jsonParser;
+    }
+
+    // ------------------- LOGIC----------------------------
+    
+    @Override
+    public String getPatentMetadata(ImportedPatent patent) throws Exception {
+        return getPatentMetadata(patent, getSecurityToken(), 0);
+    }
+
+    // ------------------- PRIVATE -------------------------
+    
+    private String getPatentMetadata(ImportedPatent patent, String securityToken, int retryCount) throws Exception {
+        
+        if (retryCount > maxRetriesCount) {
+            throw new PatentServiceException("number of maximum retries exceeded: " + maxRetriesCount);
+        }
+
+        HttpRequest httpRequest = buildPatentMetaRequest(patent, securityToken, opsUriRoot);
+        HttpResponse httpResponse = httpClient.execute(opsHost, httpRequest);
+
+        int statusCode = httpResponse.getStatusLine().getStatusCode();
+
+        switch (statusCode) {
+        case 200: {
+            HttpEntity entity = httpResponse.getEntity();
+            if (entity != null) {
+                return EntityUtils.toString(entity);
+            } else {
```

Comment:
Hmm, I guess entity can be null. I think it's the other way round: I should not call `httpResponse.getEntity()` when building exception message because it's null. I have fixed that.

Reply:
Yes, I missed the nullability of entity. But my comment was that `else` condition check is not necessary, we can have something like this
```java
            case 200: {
                HttpEntity entity = httpResponse.getEntity();
                if (entity != null) {
                    return EntityUtils.toString(entity);
                }
                throw new PatentServiceException(
                        "got empty entity in response, full status: " + httpResponse.getStatusLine());
```

Reply:
Matter of style, but I would rather keep the `else` in such cases. I think it better shows that the two paths are related.
_But_ if you suggested turning the condition around to
```java
// check prerequisite
if (entity == null) { throw ...; }
// prerequisite verified, do stuff
return ...;
```
then maybe it would read better without `else`.

Reply:
I went with a little bit altered version on Michal's code:
```java
if (entity == null) { throw ...; }
return ...;
```